---
title: "Guilhamon_preprocessing"
author: "Ronald Wu"
date: '2024-03-05'
output: html_document
---
# Introduction
The purpose of this notebook is to a document the process for downloading and pre-processing the original files from Paul's scATAC-seq [study](https://elifesciences.org/articles/64090). The supplementary processed files deposited on GEO as [GSE139136](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE139136) was insufficient for our analysis using Signac. Specifically, we need the fragments.tsv.gz file, which can be generated by running cellranger on the original fastq files. These raw files are available on SRA.

# Load Modules on Computing Cluster
Make sure you are on the H4H cluster login node, since we need internet connection to download the datasets.

```{bash load modules}
module load sratoolkit/3.0.7
```

# Download FASTQ-files from SRA-accessions

## Method 1: Using commands from SRA toolkit (prefetch)

The login node for the computing cluster has access to the internet. We can load the SRA toolkit module, which is installed on the cluster, and run commands to download the dataset given a SRA accession number. The expected files will be in `.sra` format and other helper files.

For more details on the commands used, see the [documentation](https://github.com/ncbi/sra-tools/wiki/08.-prefetch-and-fasterq-dump).

Note:
The prefetch downloading process is very slow and many users online prefer direct download (see method 2). However, using `prefetch` followed by `fastq-dump` or `fasterq-dump` allowed more flexibility for building the fastq files. For example, the fastq built from the SRA file may differ from the fastq from direct download in terms of their header.

```{bash download files from SRA}
# on the login node
module load sratoolkit/3.0.7
cd ~
mkdir SRA
cd ~/SRA

prefetch SRR10315835
prefetch SRR10315836
prefetch SRR10315837
prefetch SRR10315838
```
Note: The size of the .sra file (~4 gb each) may exceed allocations for the login node. Additional files (small in ziare also downloaded in the folder containing the SRA file.
> [rwu@h4huhnlogin1 SRA]$ ls SRR10315835/
CM000663.1  CM000672.1  CM000681.1  GL000194.1  GL000203.1  GL000212.1  GL000221.1  GL000230.1  GL000239.1  GL000248.1
CM000664.1  CM000673.1  CM000682.1  GL000195.1  GL000204.1  GL000213.1  GL000222.1  GL000231.1  GL000240.1  GL000249.1
CM000665.1  CM000674.1  CM000683.1  GL000196.1  GL000205.1  GL000214.1  GL000223.1  GL000232.1  GL000241.1  NC_007605.1
CM000666.1  CM000675.1  CM000684.1  GL000197.1  GL000206.1  GL000215.1  GL000224.1  GL000233.1  GL000242.1  NC_012920.1
CM000667.1  CM000676.1  CM000685.1  GL000198.1  GL000207.1  GL000216.1  GL000225.1  GL000234.1  GL000243.1  SRR10315835.sra
CM000668.1  CM000677.1  CM000686.1  GL000199.1  GL000208.1  GL000217.1  GL000226.1  GL000235.1  GL000244.1  SRR10315835.sra.vdbcache
CM000669.1  CM000678.1  GL000191.1  GL000200.1  GL000209.1  GL000218.1  GL000227.1  GL000236.1  GL000245.1
CM000670.1  CM000679.1  GL000192.1  GL000201.1  GL000210.1  GL000219.1  GL000228.1  GL000237.1  GL000246.1
CM000671.1  CM000680.1  GL000193.1  GL000202.1  GL000211.1  GL000220.1  GL000229.1  GL000238.1  GL000247.1

```{bash extract with fastq-dump on login node}
# on login node (internet enabled)
module load sratoolkit/3.0.7

cd /cluster/home/rwu/SRA
fastq-dump --split-files --origfmt --defline-seq '@rd.$si:$sg:$sn' SRR10315835

```
Above `fastq-dump` command from [Not A Rocket Scientist](https://notarocketscientist.xyz/posts/2022-07-26-getting-index-reads-from-sra/).

Explanation of flags used:
1. `--split-files`: Dump each read into separate file. Files will receive suffix corresponding to read number.
2. `--origfmt`: Defline contains only original sequence name
3. `--defline-seq` '@rd.$si:$sg:$sn' : Defline format specification for sequence.
  + where $si is spot id, $sg is spot group, $sn is spot name
  
See this [Edward's lab blog](https://edwards.flinders.edu.au/fastq-dump/) for even more descriptions.

Needs to be done on build partition since you will get a `fail to call services` error if no internet. However, we ran out of space here, so the prefetched `SRR10315835` folder was moved to the compute node. See below for attempt on compute node (after disabling internet).

```{bash extract with fastq-dump on compute node (without internet)}
module load sratoolkit/3.0.7

# run vdb-config interactively and disable remote access (internet)
vdb-config -i

cd /cluster/projects/wouterslab/scATAC/Guilhamon/SRA

# added flags for output directory and verbose
fastq-dump --split-files --origfmt --defline-seq '@rd.$si:$sg:$sn' SRR10315835 -v --outdir fastq/
cd fastq
ls -ltrh
```

> [rwu@node124 fastq]$ ls -ltrh
total 60G
-rw-r--r-- 1 rwu wouterslab 25G Mar 11 18:15 SRR10315835_1.fastq
-rw-r--r-- 1 rwu wouterslab 25G Mar 11 18:15 SRR10315835_2.fastq
> rd.1:CTTTCCTT:1
GGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTATG
+1
AAAAA/EAA/A/////EEA//EEE6A/EEEA//AAEA//EEE///EEE//
[rwu@node116 fastq]$ cat GBM_4218_S1_L001_R2_001.fastq | head -n 4
@rd.1:CTTTCCTT:1
AACCCTAACACTAAACATAAACCAAAACCTAAAACTAACAATAAACCTAA
+1
AA/A//EE//E/EA///EEE/<A/AE/EE//E//E/EEA//6/E//<//E

The reads from read1 and read2 are 50 bp as expected.

> [rwu@node116 fastq]$ awk 'NR%4==1' GBM_4218_S1_L001_R2_001.fastq | head -5
@rd.1:CTTTCCTT:1
@rd.2:ACCAGTCC:2
@rd.3:TAGGTAAA:3
@rd.4:GGACAGGG:4
@rd.5:TAGGTAAA:5

These look like i7 8 bp sample index. 

The ~4.0G .sra file turned into 2 fastq at 50G total. When compared to the direct download of fastq (method 2), this is significantly larger but here only has 2 files vs 3 fastq.gz files.

```{bash edward's lab commands}
# Try commands from Edwards's lab (from 2015):
# fastq-dump --gzip --skip-technical  --readids --dumpbase --split-files --clip SRR10315835

# Try commands from Edwards's lab (from 2020):
fastq-dump --outdir fastq --skip-technical  --readids --read-filter pass --dumpbase --clip --split-3 SRR10315835
# Rejected 367 READS because of filtering out non-biological READS
# Read 187230591 spots for SRR10315835
# Written 80107509 spots for SRR10315835
```
>
[rwu@node131 fastq]$ pwd
/cluster/projects/wouterslab/scATAC/Guilhamon/SRA/fastq
[rwu@node131 fastq]$ ls -ltrh
total 34G
-rw-r--r-- 1 rwu wouterslab 14G Mar 13 10:09 SRR10315835_pass_1.fastq
-rw-r--r-- 1 rwu wouterslab 14G Mar 13 10:09 SRR10315835_pass_2.fastq
-rw-r--r-- 1 rwu wouterslab 55K Mar 13 10:09 SRR10315835_pass.fastq
[rwu@node131 fastq]$ head SRR10315835_pass_1.fastq
@SRR10315835.1.1 1 length=50
GGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTATG
+SRR10315835.1.1 1 length=50
AAAAA/EAA/A/////EEA//EEE6A/EEEA//AAEA//EEE///EEE//
@SRR10315835.4.1 4 length=50
AACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAC
+SRR10315835.4.1 4 length=50
AAAAAEEAEEEAEAEEEEEEEEEEE6EE//EE//EE66A/EAE//E/A//
@SRR10315835.6.1 6 length=46
AACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACC
[rwu@node131 fastq]$ head SRR10315835_pass_2.fastq
@SRR10315835.1.2 1 length=50
AACCCTAACACTAAACATAAACCAAAACCTAAAACTAACAATAAACCTAA
+SRR10315835.1.2 1 length=50
AA/A//EE//E/EA///EEE/<A/AE/EE//E//E/EEA//6/E//<//E
@SRR10315835.4.2 4 length=50
GGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTATG
+SRR10315835.4.2 4 length=50
AAAAA6EEEEE6EEEEEEEEAEE6AEAAEAEA/EAE/E/EEE/E//EE/<
@SRR10315835.6.2 6 length=50
GGTTAGGGTTAGGGTTAGTGTTAGGGATAGGGTAAGGGTTAGGGTTATGT
[rwu@node131 fastq]$ head SRR10315835_pass.fastq
@SRR10315835.187230225.2 187230225 length=26
GGCTCGGAGATGTGTATAAGAGACAG
+SRR10315835.187230225.2 187230225 length=26
AAAAAEEEEEEEAEEEAEEEEEEEEE
@SRR10315835.187230226.1 187230226 length=23
CGTCAGATGTGTATAAGAGACAG
+SRR10315835.187230226.1 187230226 length=23
AAAA6E6EEEEE6EEEE6E6E//
@SRR10315835.187230227.2 187230227 length=50
CTGTATCTAATACACATATGAAGCTGCCGACCATCATACACTGTCCCAGA


Next, let's rename the fastq so they are in `bcl2fastq` format for use with cellranger-atac as mentioned on the 10x website:

[Sample Name]_S1_L00[Lane Number]_[Read Type]_001.fastq.gz
<<<<<<< HEAD

Where Read Type is one of:

=======

Where Read Type is one of:

>>>>>>> e1740f9b061399bf9fd7eb853883895c292b9c7a
I1: Dual index i7 read (optional)
R1: Read 1
R2: Dual index i5 read
R3: Read 2

Alternatively, Cell Ranger ATAC will also accept ATAC FASTQs in this format:

I1: Dual index i7 read (optional)
R1: Read 1
I2: Dual index i5 read
R2: Read 2

```{bash rename and run cellranger-atac}
cd /cluster/projects/wouterslab/scATAC/Guilhamon/SRA/fastq
mv SRR10315835_1.fastq GBM_4218_S1_L001_R1_001.fastq
mv SRR10315835_2.fastq GBM_4218_S1_L001_R2_001.fastq

cd /cluster/projects/wouterslab/scATAC/Guilhamon/SRA/
sbatch ~/scripts/cellranger-atac-count.sh \
GBM_4218 \
fastq/
```
We got an error:
> 
Unable to read barcode sequence for read ID rd.1:CTTTCCTT:1: there was no I2 read FASTQ and we were unable to read a 16-base barcode from the FASTQ header. Make sure that the flow cell was demultiplexed correctly.
Waiting 6 seconds for UI to do final refresh.
Pipestance failed. Use --noexit option to keep UI running after failure. 
2024-03-12 00:31:42 Shutting down.

Looks like they are looking for a 16-base barcode.

## Method 1b: fasterq-dump
```{bash extract with fasterq-dump}
module load sratoolkit/3.0.7
cd /cluster/projects/wouterslab/scATAC/Guilhamon/SRA

#fasterq-dump --split-files --include-technical SRR10315835 --outdir fastq2
fasterq-dump --split-files --include-technical SRR10315835 --outdir fastq2 --verbose

# try --split-3 flag
fasterq-dump --split-3 --include-technical SRR10315835 --outdir fastq2 --verbose

# try another accession
fasterq-dump --split-files --include-technical SRR10315836 --outdir fastq2 --verbose
fasterq-dump --split-files --skip-technical SRR10315836 --outdir fastq2 --verbose

```
Instructions for fasterq-dump can be found [here](https://github.com/ncbi/sra-tools/wiki/HowTo:-fasterq-dump).
fasterq-dump is faster than fastq-dump, but the syntax is a bit different. Can also use without internet if specified using `vdb-config -i`. 

Ran very fast compared to `fastq-dump`, but problems:
1. No fastq files were made in the fastq2/ outdir with either `--split-files` or `--split-3` flag.
2. The `fasterq-tmp` folder was generated with `.dat` files but was not removed. Usually this temp folder is removed and fastq is generated. 
3. Didn't see any errors when we ran the command with `--verbose`.

## Bare accession needs internet
```{bash fasterq-dump SRR000001}
cd /cluster/home/rwu/SRA

# run vdb-config interactively and enable remote access (internet)
vdb-config -i

# default (split-3 is used by default, explicit flag is --split-3)
fasterq-dump SRR000001
```
> spots read      : 470,985
reads read      : 1,883,940
reads written   : 707,026
reads 0-length  : 468,635
technical reads : 708,279
> [rwu@h4huhnlogin1 SRA]$ ls -ltrh
total 299M
-rw-r--r-- 1 rwu wouterslab  584 Mar  6 11:53 paul.txt
drwxr-xr-x 2 rwu wouterslab 2.4K Mar  8 21:04 SRR10315835
-rw-r--r-- 1 rwu wouterslab  68M Mar 11 16:17 SRR000001_1.fastq
-rw-r--r-- 1 rwu wouterslab  76M Mar 11 16:17 SRR000001_2.fastq
-rw-r--r-- 1 rwu wouterslab 141M Mar 11 16:17 SRR000001.fastq

The temp folder was removed and we ended up with 3 fastq for SRR000001 in the current working directory
Note: On compute cluster, the same command returned an error
> fasterq-dump quit with error code 3


## Method 2: Using custom script from sra-explorer

There is a convenient [website](https://sra-explorer.info/) that let's you search for all the datasets you want to download and adds them to a shopping cart. Then, this cart will show and format various methods for download, including bash scripts which we can use on the cluster.

All 4 of the scATAC-seq samples were entered:
1. SRR10315835
2. SRR10315836
3. SRR10315837
4. SRR10315838

The following bash script was generated to download the fastq files directly. Note: The header was changed for use on the slurm scheduler. The build partition is selected because it has internet access.

```{bash download files via script}
#!/bin/sh
#SBATCH -t 2:00:00
#SBATCH -J sradownload
#SBATCH -p build
#SBATCH -c 1
#SBATCH -N 1
#SBATCH -o %x-%j.out
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR103/035/SRR10315835/SRR10315835.fastq.gz -o SRR10315835_GSM4131776_GBM_4218_Homo_sapiens_ATAC-seq.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR103/035/SRR10315835/SRR10315835_1.fastq.gz -o SRR10315835_GSM4131776_GBM_4218_Homo_sapiens_ATAC-seq_1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR103/035/SRR10315835/SRR10315835_2.fastq.gz -o SRR10315835_GSM4131776_GBM_4218_Homo_sapiens_ATAC-seq_2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR103/038/SRR10315838/SRR10315838.fastq.gz -o SRR10315838_GSM4131779_GBM_4349_Homo_sapiens_ATAC-seq.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR103/038/SRR10315838/SRR10315838_1.fastq.gz -o SRR10315838_GSM4131779_GBM_4349_Homo_sapiens_ATAC-seq_1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR103/038/SRR10315838/SRR10315838_2.fastq.gz -o SRR10315838_GSM4131779_GBM_4349_Homo_sapiens_ATAC-seq_2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR103/036/SRR10315836/SRR10315836.fastq.gz -o SRR10315836_GSM4131777_GBM_4250_Homo_sapiens_ATAC-seq.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR103/036/SRR10315836/SRR10315836_1.fastq.gz -o SRR10315836_GSM4131777_GBM_4250_Homo_sapiens_ATAC-seq_1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR103/036/SRR10315836/SRR10315836_2.fastq.gz -o SRR10315836_GSM4131777_GBM_4250_Homo_sapiens_ATAC-seq_2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR103/037/SRR10315837/SRR10315837.fastq.gz -o SRR10315837_GSM4131778_GBM_4275_Homo_sapiens_ATAC-seq.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR103/037/SRR10315837/SRR10315837_1.fastq.gz -o SRR10315837_GSM4131778_GBM_4275_Homo_sapiens_ATAC-seq_1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR103/037/SRR10315837/SRR10315837_2.fastq.gz -o SRR10315837_GSM4131778_GBM_4275_Homo_sapiens_ATAC-seq_2.fastq.gz
```

The above script was saved as `paul.txt`. 
Each sample has three files, two of which are the paired end reads denoted with _1 and _2 suffix. 
The third file is small and can be ignored according to [this biostars post](https://www.biostars.org/p/156909/). They are apparently orphaned reads after trimming.

Also, the build partition has a 2 hr time limit, which might not be enough to download all the files.
```{bash run script}
cd ~/SRA
squeue paul.txt
```

```{bash copy files over}
cd /cluster/projects/wouterslab
mkdir -p scATAC/Guilhamon

cp ~/SRA/*fastq.gz /cluster/projects/wouterslab/scATAC/Guilhamon
```

## Method 3:
Download the original BAM files and then use `bamtofastq` to convert to fastq. 

# Rename fastq for use with Cell Ranger

Cellranger requires the fastq files to follow the naming convention employed by `bcl2fastq`:

[Sample Name]_S1_L00[Lane Number]_[Read Type]_001.fastq.gz

Where Read Type is one of:

I1: Sample index read (optional)
I2: Sample index read (optional)
R1: Read 1 (required)
R2: Read 2 (required)

```{bash renaming fastqs}
cd /cluster/projects/wouterslab/scATAC/Guilhamon
# make folders named after each patient sample
mkdir fastq/{GBM_4218,GBM_4250,GBM_4275,GBM_4349}
# move fastqs
mv *GBM_4218*_{1,2}* fastq/GBM_4218/
mv *GBM_4250*_{1,2}* fastq/GBM_4250/
mv *GBM_4275*_{1,2}* fastq/GBM_4275/
mv *GBM_4349*_{1,2}* fastq/GBM_4349/

# rename files
mv fastq/GBM_4218/SRR10315835_GSM4131776_GBM_4218_Homo_sapiens_ATAC-seq_1.fastq.gz fastq/GBM_4218/GBM_4218_S1_L001_R1_001.fastq.gz
mv fastq/GBM_4218/SRR10315835_GSM4131776_GBM_4218_Homo_sapiens_ATAC-seq_2.fastq.gz fastq/GBM_4218/GBM_4218_S1_L001_R2_001.fastq.gz

mv fastq/GBM_4250/SRR10315836_GSM4131777_GBM_4250_Homo_sapiens_ATAC-seq_1.fastq.gz fastq/GBM_4250/GBM_4250_S1_L001_R1_001.fastq.gz
mv fastq/GBM_4250/SRR10315836_GSM4131777_GBM_4250_Homo_sapiens_ATAC-seq_2.fastq.gz fastq/GBM_4250/GBM_4250_S1_L001_R2_001.fastq.gz

mv fastq/GBM_4275/SRR10315837_GSM4131778_GBM_4275_Homo_sapiens_ATAC-seq_1.fastq.gz fastq/GBM_4275/GBM_4275_S1_L001_R1_001.fastq.gz
mv fastq/GBM_4275/SRR10315837_GSM4131778_GBM_4275_Homo_sapiens_ATAC-seq_2.fastq.gz fastq/GBM_4275/GBM_4275_S1_L001_R2_001.fastq.gz

mv fastq/GBM_4349/SRR10315838_GSM4131779_GBM_4349_Homo_sapiens_ATAC-seq_1.fastq.gz fastq/GBM_4349/GBM_4349_S1_L001_R1_001.fastq.gz
mv fastq/GBM_4349/SRR10315838_GSM4131779_GBM_4349_Homo_sapiens_ATAC-seq_2.fastq.gz fastq/GBM_4349/GBM_4349_S1_L001_R2_001.fastq.gz

```

However, the files we download from SRA are not named in that format and require some modification
# Run cellranger-atac count pipeline
Since we are working with fastq files directly, we can skip the part of the pipeline where the raw bcl files are converted to fastq files using `cellranger-atac mkfastq`. We can directly use the `cellranger-atac count` function.

The bottom is example code for running `cellranger-atac count` locally from the 10x genomics website:

```{bash run cellranger-atac count locally}
cd /home/jdoe/runs
cellranger-atac count --id=sample345 \
                        --reference=/opt/refdata-cellranger-arc-GRCh38-2020-A-2.0.0 \
                        --fastqs=/home/jdoe/runs/HAWT7ADXX/outs/fastq_path \
                        --sample=mysample \
                        --localcores=8 \
                        --localmem=64
```

Some questions based on the example commands above:
1. Where is the transcriptome reference on the H4h cluster?
There are several options:
1. For cell ranger atac version 2.0.0
/cluster/tools/data/commondata/cellranger/refdata-cellranger-arc-GRCh38-2020-A-2.0.0
2. For use with cell ranger atac version 1.2.0
/cluster/tools/data/commondata/cellranger/refdata-cellranger-atac-GRCh38-1.2.0
3. for gene expression (GEX) analysis
/cluster/tools/data/commondata/cellranger/refdata-gex-GRCh38-2020-A

Let's use `cellranger/refdata-cellranger-arc-GRCh38-2020-A-2.0.0`, since it was the one used in the 10x genomics atac website. Also, `cellranger-atac/2.1.0` is the module on the cluster.

Next, we need to modify this code and add it to a script to submit as a batch job on slurm. Some examples of this can be found on [biowolf](https://hpc.nih.gov/apps/cellranger-atac.html). We saved the script `cellranger-atac-count.sh` at the ~/scripts folder.

```{bash cellranger-atac-count.sh batch job }
cd /cluster/projects/wouterslab/scATAC/Guilhamon

sbatch ~/scripts/cellranger-atac-count.sh \
GBM_4218 \
fastq/GBM_4218

sbatch ~/scripts/cellranger-atac-count.sh \
GBM_4250 \
fastq/GBM_4250

sbatch ~/scripts/cellranger-atac-count.sh \
GBM_4275 \
fastq/GBM_4275

sbatch ~/scripts/cellranger-atac-count.sh \
GBM_4349 \
fastq/GBM_4349
```

Troubleshooting:
1. Remember we had three fastq files downloaded. perhaps the third one is the i5 index. let's rename it
apparently R3 is used to denote read 2
```
mv GBM_4218_S1_L001_R2_001.fastq.gz GBM_4218_S1_L001_R3_001.fastq.gz
mv SRR10315835_GSM4131776_GBM_4218_Homo_sapiens_ATAC-seq.fastq.gz fastq/GBM_4218
mv SRR10315835_GSM4131776_GBM_4218_Homo_sapiens_ATAC-seq.fastq.gz GBM_4218_S1_L001_R2_001.fastq.gz
```
