---
title: "Guilhamon"
author: "Ronald Wu"
date: '2024-01-18'
output: html_document
---

# Install Missing libraries

```{r install missing libs, echo = FALSE, eval = FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("EnsDb.Hsapiens.v75")
```

# Load libraries
```{r load libraries}
library(Signac) # 1.12.9003 developmental
library(Seurat)
library(EnsDb.Hsapiens.v75) # hg19
library(biovizBase)
library(hdf5r)
# library(ggplot2)
library(tidyverse)
library(patchwork)

library(Matrix)
library(irlba)
```

# Downloading Guilhamon et al scATAC-seq data
The associated data is from GEO [GSE139136](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE139136). 

There are four patients:
+ GSM4131776	GBM_4218
+ GSM4131777	GBM_4250
+ GSM4131778	GBM_4275
+ GSM4131779	GBM_4349

## Data processing:
All raw base call (BCL) files generated by Illumina® sequencers were converted to FASTQ files using cellranger-atac mkfastq function, a wrapper around bcl2fastq from Illumina®, with additional 10x Genomics-specific features. Next, read filtering and alignment, barcode counting, and identification of transposase cut sites were detected using cellranger-atac count function using the cellranger-atac-specific hg19 reference package.
Genome_build: refdata-cellranger-atac-hg19-1.1.0 (hg19 with pre-generated indices and other data required to run CellRanger-ATAC)
Supplementary_files_format_and_content: Filtered peak-barcode matrix output from CellRanger-ATAC (containing only detected cellular barcodes)


We will download the following 4 files:
1. Peak by cell matrix (filtered)
2. Per Barcode metrics (CSV)
3. Fragments (TSV)
4. Fragments index (TBI)

We will be working on the output data from Cell Ranger 1.1.0. 
+ Sample: Fresh Cortex from Adult Mouse brain (P50)
+ Output: Cell Ranger ATAC 1.1.0
+ ~7700 transposed nuclei were loaded and 5337 nuclei were recovered
The associated data for this vignette is from [10x genomics](https://www.10xgenomics.com/resources/datasets/fresh-cortex-from-adult-mouse-brain-p-50-1-standard-1-1-0). 


```{bash download all 4 patient data from GSE139136, echo = FALSE, eval = FALSE}
# GSM4131776
wget -P 'data/Guilhamon/GSM4131776/' \
https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4131nnn/GSM4131776/suppl/GSM4131776%5F4218%5Fmatrix.mtx.gz

wget -P 'data/Guilhamon/GSM4131776/' \
https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4131nnn/GSM4131776/suppl/GSM4131776%5F4218%5Fbarcodes.tsv.gz

wget -P 'data/Guilhamon/GSM4131776/' \
https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4131nnn/GSM4131776/suppl/GSM4131776%5F4218%5Fpeaks.bed.gz

# GSM4131777
wget -P 'data/Guilhamon/GSM4131777/' \
https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4131nnn/GSM4131777/suppl/GSM4131777%5F4250%5Fbarcodes.tsv.gz

wget -P 'data/Guilhamon/GSM4131777/' \
https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4131nnn/GSM4131777/suppl/GSM4131777%5F4250%5Fmatrix.mtx.gz

wget -P 'data/Guilhamon/GSM4131777/' \
https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4131nnn/GSM4131777/suppl/GSM4131777%5F4250%5Fpeaks.bed.gz

# GSM4131778
wget -P 'data/Guilhamon/GSM4131778/' \
https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4131nnn/GSM4131778/suppl/GSM4131778%5F4275%5Fbarcodes.tsv.gz

wget -P 'data/Guilhamon/GSM4131778/' \
https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4131nnn/GSM4131778/suppl/GSM4131778%5F4275%5Fmatrix.mtx.gz

wget -P 'data/Guilhamon/GSM4131778/' \
https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4131nnn/GSM4131778/suppl/GSM4131778%5F4275%5Fpeaks.bed.gz

# GSM4131779
wget -P 'data/Guilhamon/GSM4131779/' \
https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4131nnn/GSM4131779/suppl/GSM4131779%5F4349%5Fbarcodes.tsv.gz

wget -P 'data/Guilhamon/GSM4131779/' \
https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4131nnn/GSM4131779/suppl/GSM4131779%5F4349%5Fmatrix.mtx.gz

wget -P 'data/Guilhamon/GSM4131779/' \
https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4131nnn/GSM4131779/suppl/GSM4131779%5F4349%5Fpeaks.bed.gz
```

There are at least two options to load in data downloaded from GEO, either using `Read10x()` or `ReadMtx()`. Since the files downloaded from GEO have a prefix with their sample name, ReadMtx is preferred to read in those files, as you can just supply their path. In contrast, Read10x expects non-prefixed files from Cell Ranger output.

For ReadMtx(), there are three params required:
1. mtx: this is the matrix file, matrix.mtx.gz
2. cells: this is the barcodes file
3. features: this is the features.tsv.gz

Note: For GSE139136, each patient sample has three files, such as the matrix, barcodes, and peaks.bed file. The features or fragments file is not provided.

We need a `fragments.tsv.gz` for the `Signac::CreateChromatinAssay()` function. This is the Barcoded and aligned fragment file.

Other people have also encountered this issue where downloading a dataset from GEO, a features file is [missing](https://github.com/satijalab/seurat/issues/6270). However, the barcodes, matrix files, and peaks.bed are provided. A proposed solution is to create one using awk from the bed file, but no examples are provided or could be found on the internet.

Interestingly, the PBMC Signac vignette mentions a way to create the Seurat object in absence of the fragments file, which I have implemented below:
```{r load object}
# For output from CellRanger < 3.0
  #data_dir <- 'data/Guilhamon/GSM4131776/'
  #list.files(data_dir) # Should show barcodes.tsv, genes.tsv, and matrix.mtx
  # expression_matrix <- Read10X(data.dir = data_dir)

#expression_matrix <- ReadMtx(
#  mtx = "data/Guilhamon/GSM4131776/GSM4131776_4218_matrix.mtx.gz", 
#  cells = "data/Guilhamon/GSM4131776/GSM4131776_4218_barcodes.tsv.gz",
#  features = "data/Guilhamon/GSM4131776/GSM4131776_4218_peaks.bed.gz",
#  feature.column = 1, skip.cell = 0, skip.feature = 0
#)

# If missing .h5 file, but have matrix, barcodes, peaks (from Signac PBMC vignette): 
counts <- Matrix::readMM("data/Guilhamon/GSM4131776/GSM4131776_4218_matrix.mtx.gz")
barcodes <- readLines("data/Guilhamon/GSM4131776/GSM4131776_4218_barcodes.tsv.gz")
peaks <- read.table("data/Guilhamon/GSM4131776/GSM4131776_4218_peaks.bed.gz", sep="\t")
peaknames <- paste(peaks$V1, peaks$V2, peaks$V3, sep="-")

colnames(counts) <- barcodes
rownames(counts) <- peaknames

# create ChromatinAssay Object
G4218_chr_assay <- CreateChromatinAssay(
  counts = counts,
  sep = c(":", "-"),
  genome = "hg19",
#  fragments = 'data/vignette_data/atac_v1_adult_brain_fresh_5k_fragments.tsv.gz',
  min.cells = 1
)

# create Seurat Object
G4218 <- CreateSeuratObject(
  counts = G4218_chr_assay,
  assay = 'peaks',
  project = 'ATAC',
  # meta.data = metadata
)

# remove chromatin assay from memory
G4218_chr_assay <- NULL
```
> G4218
An object of class Seurat 
112164 features across 1600 samples within 1 assay 
Active assay: peaks (112164 features, 0 variable features)
 2 layers present: counts, data
 
Note: we are missing two files
+ the fragments file fragments.tsv.gz
+ metadata file

# Add gene annotations

```{r gene annotations}
# extract gene annotations from EnsDb
annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v75)

# change to UCSC style since the data was mapped to hg19
seqlevels(annotations) <- paste0('chr', seqlevels(annotations))
genome(annotations) <- "hg19"
# add the gene information to the object
Annotation(G4218) <- annotations
```

# QC Metrics
skipped because no fragments file

```{R plot QC metrics}
VlnPlot(
  object = G4218,
  features = c('nCount_peaks', 
              # 'TSS.enrichment', 
               #'blacklist_ratio', 
               #'nucleosome_signal', 
               #'pct_reads_in_peaks',
               'nFeature_peaks'),
  pt.size = 0.1,
  ncol = 5
)
```

# Subset based on QC metrics
```{r subset}
G4218_subset <- subset(
  x = G4218,
    subset = nCount_peaks > 3000 &
    nCount_peaks < 30000
    # pct_reads_in_peaks > 15 &
    # blacklist_ratio < 0.05 &
    # nucleosome_signal < 4 &
    # TSS.enrichment > 3
)
G4218_subset

# how did the paper get 1433 cells?
(G4218$nCount_peaks) > 2000 %>% nrow()
```
> G4218_subset
An object of class Seurat 
112164 features across 647 samples within 1 assay 
Active assay: peaks (112164 features, 0 variable features)
 2 layers present: counts, data
 
After QC filtering: went from 1600 to 647 cells. Is this correct? Let's see their published paper: In figure 1 A, the umap says 1433 cells. Means that they did some filtering, but methods doesn't indicate any filtering params.

Note, many of the QC metrics were not available due to two reasons:
1. no metadata file
2. no fragments file
In the future, should revisit and get these files by running cell ranger on the raw data.

# Normalization and linear dimensional reduction
```{r Normalization and linear dimensional reduction}
G4218_subset <- RunTFIDF(G4218_subset)
G4218_subset <- FindTopFeatures(G4218_subset, min.cutoff = 'q0')
G4218_subset <- RunSVD(G4218_subset)
```

Strong correlation with the first component, so we will remember to remove it downstream.
```{r export pdf, eval=FALSE}
pdf("results/Guilhamon/2024-01-30_G4218_DepthCor_plot.pdf")
DepthCor(sf519_5)
dev.off()
```
