---
title: "scATAC"
author: "Ronald Wu"
date: '2024-01-08'
output: html_document
---
# Single Cell ATAC-seq Analysis
There are a few well known packages in R and python that can be used to analyze scATAC-seq data.
+ R Packages:
  + Signac: developed by Tim Stuart while in the Satija lab. Actively maintained by Stuart lab.
    + [website](https://stuartlab.org/signac/)
  + ArchR: 
    + [website](https://www.archrproject.com/)

# Signac Vignettes
The purpose of this notebook is to learn how to use `Signac` for processing and analyzing scATAC-seq data. We'll be following the vignettes provided by the Stuart lab, which is the original developer of this package. All code is slightly modified from their original vignettes from the following:
+ [mouse_brain_vignette](https://stuartlab.org/signac/articles/mouse_brain_vignette)
+ 

# Setup
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Install packages
```{r install packages}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("EnsDb.Mmusculus.v79")
BiocManager::install("biovizBase")

remotes::install_github("stuart-lab/signac", ref="develop") # installs v1.12.9003

install.packages("Matrix", type = "source")
install.packages("irlba", type = "source")

require(remotes)
#install_version("Matrix", version = "1.5-3", repos = "http://cran.us.r-project.org")
#install_version("Matrix", version = "1.6-1.1", repos = "http://cran.us.r-project.org")
#install_version("Matrix", version = "1.6.3", repos = "http://cran.us.r-project.org")

#install_version("SeuratObject", version = "4.1.4", repos = "http://cran.us.r-project.org")

```

# Load Libraries
```{r load libraries}
library(Signac) # 1.12.9003 developmental
library(Seurat)
library(EnsDb.Mmusculus.v79)
library(biovizBase)
library(hdf5r)
# library(ggplot2)
library(tidyverse)
library(patchwork)

library(Matrix)
library(irlba)
```

# Downloading 10x genomics scRNA-seq Adult mouse brain
The associated data for this vignette is from [10x genomics](https://www.10xgenomics.com/resources/datasets/fresh-cortex-from-adult-mouse-brain-p-50-1-standard-1-1-0). 

We will download the following 4 files:
1. Peak by cell matrix (filtered)
2. Per Barcode metrics (CSV)
3. Fragments (TSV)
4. Fragments index (TBI)

We will be working on the output data from Cell Ranger 1.1.0. 
+ Sample: Fresh Cortex from Adult Mouse brain (P50)
+ Output: Cell Ranger ATAC 1.1.0
+ ~7700 transposed nuclei were loaded and 5337 nuclei were recovered


```{bash download 10x data; echo = FALSE}
# Peak by cell matrix HDF5 (filtered)
wget -P 'data/vignette_data' \
http://cf.10xgenomics.com/samples/cell-atac/1.1.0/atac_v1_adult_brain_fresh_5k/atac_v1_adult_brain_fresh_5k_filtered_peak_bc_matrix.h5

# Per Barcode metrics (CSV)
wget -P 'data/vignette_data' \
http://cf.10xgenomics.com/samples/cell-atac/1.1.0/atac_v1_adult_brain_fresh_5k/atac_v1_adult_brain_fresh_5k_singlecell.csv

# Fragments (TSV)
wget -P 'data/vignette_data' \
http://cf.10xgenomics.com/samples/cell-atac/1.1.0/atac_v1_adult_brain_fresh_5k/atac_v1_adult_brain_fresh_5k_fragments.tsv.gz

# Fragments index (TBI)
wget -P 'data/vignette_data' \
http://cf.10xgenomics.com/samples/cell-atac/1.1.0/atac_v1_adult_brain_fresh_5k/atac_v1_adult_brain_fresh_5k_fragments.tsv.gz.tbi

```

# Pre-processing

```{r pre-processing; echo = FALSE}

# load counts matrix and metadata
counts <- Read10X_h5("data/vignette_data/atac_v1_adult_brain_fresh_5k_filtered_peak_bc_matrix.h5")
metadata <- read.csv(
  file = "data/vignette_data/atac_v1_adult_brain_fresh_5k_singlecell.csv",
  header = TRUE,
  row.names = 1
)

# create ChromatinAssay Object
brain_assay <- CreateChromatinAssay(
  counts = counts,
  sep = c(":", "-"),
  genome = "mm10",
  fragments = 'data/vignette_data/atac_v1_adult_brain_fresh_5k_fragments.tsv.gz',
  min.cells = 1
)

# create Seurat Object
brain <- CreateSeuratObject(
  counts = brain_assay,
  assay = 'peaks',
  project = 'ATAC',
  meta.data = metadata
)

# extract gene annotations from EnsDb
annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Mmusculus.v79)

# change to UCSC style since the data was mapped to GRCm38
seqlevels(annotations) <- paste0('chr', seqlevels(annotations))
genome(annotations) <- "mm10"

# add the gene information to the object
Annotation(brain) <- annotations
```

# QC Metrics
## Nucleosome Signal and Banding Patterns
Let's look at the following QC metrics to ensure good data quality prior to further analysis:
+ Nucleosome signal and banding pattern: A low nucleosome signal, which is the ratio of mono-nucleosome to nucleosome-free fragments, is preferred. We should also see a strong nucleosome banding pattern.  
+ Transcriptional start site (TSS) enrichment score: ratio of fragments at the TSS vs TSS-flanking regions for each cell. This is an indicator of signal-to-noise.

```{r QC: Nucleosome Signal and banding pattern}
# calculates nucleosome signal score per cell
brain <- NucleosomeSignal(object = brain)

# group cells by nucleosome signal
brain$nucleosome_group <- ifelse(brain$nucleosome_signal > 4, 'NS > 4', 'NS < 4')

# how many cells are NS < 4
sum((brain$nucleosome_group == 'NS < 4')) # 5252

# plot nucleosome signal
hist(brain$nucleosome_signal)

# plot FragmentHistogram at chr 1
FragmentHistogram(object = brain,region = 'chr1-1-10000000')
# plot FragmentHistogram grouped by nucleosome signal at chr 1
FragmentHistogram(object = brain, group.by = 'nucleosome_group', region = 'chr1-1-10000000')

```
## TSS Enrichment Score

```{r TSS Enrichment Score; echo = FALSE}
# calculate TSS enrichment score 
# note: To plot enrichment score centered at TSS in later steps, set FAST = FALSE (matrix of integration counts at each site required (slow)))
# otherwise, set fast = TRUE if want to be faster, but won't be able to generate TSS centered plot

#brain <- TSSEnrichment(brain, fast = TRUE)
brain <- TSSEnrichment(brain, fast = FALSE)

# histogram of TSS enrichment scores
as_tibble(brain$TSS.enrichment) %>%
  ggplot(aes(x = value)) +
  geom_histogram() +
  geom_vline(xintercept = 2,
             linetype = "dotted",
             color = "red")

# group/categorize TSS enrichment as high or low
brain$high.tss <- ifelse(brain$TSS.enrichment > 2, 'High', 'Low')

# plot TSS enrichment score (centered at TSS)
TSSPlot(brain, group.by = 'high.tss') + Seurat::NoLegend()

```
Dealing with an error for TSSEnrichment( fast = FALSE)
> Error in slot(object = object, name = layer) : 
  argument "slot" is missing, with no default
  
This issue has been reported and solved on [github](https://github.com/stuart-lab/signac/issues/1538). Re-install the developmental version of Signac from Github (1.12.9003). Works after restarting RStudio!

## Additional QC Metrics: read in peaks, blacklist ratio
```{r additional QC metrics}
# add fraction of reads in peaks and blacklist ratio
brain$pct_reads_in_peaks <- brain$peak_region_fragments / brain$passed_filters * 100
brain$blacklist_ratio <- brain$blacklist_region_fragments / brain$peak_region_fragments

VlnPlot(
  object = brain,
  features = c('pct_reads_in_peaks', 'peak_region_fragments',
               'TSS.enrichment', 'blacklist_ratio', 'nucleosome_signal'),
  pt.size = 0.1,
  ncol = 5
)

DensityScatter(brain, x = 'peak_region_fragments', y = 'pct_reads_in_peaks', log_x = TRUE, quantiles = TRUE)
```

## Remove outliers based on QC metrics

```{r remove outliers}
brain <- subset(
  x = brain,
  subset = peak_region_fragments > 3000 &
    peak_region_fragments < 100000 &
    pct_reads_in_peaks > 40 &
    blacklist_ratio < 0.025 &
    nucleosome_signal < 4 &
    TSS.enrichment > 2
)
brain
```

After filtering away the outliers based on the above thresholds, we see a reduction in the number of cells contained in the `brain` seurat object:
+ cell count prior to QC: 5337
+ cell count after QC: 3512 

# Normalization and linear dimension reduction
```{r normalization and linear dimension reduction}
# Term Frequency-Inverse Document Frequency Normalization (TF-IDF)
brain <- RunTFIDF(brain)
# feature selection ('q0' uses all features; 'q75' uses top 25% of all peaks (faster))
brain <- FindTopFeatures(brain, min.cutoff = 'q0')
# dimension reduction using Singular Value Decomposition
brain <- RunSVD(object = brain)
# Plot Sequencing Depth Correlation vs each reduced dimension component 
DepthCor(brain, n = NULL) # n = NULL plots all the components; default n = 10;
```
Encountered an error with Signac::RunSVD()
> Error in irlba(A = t(x = object), nv = n, work = irlba.work, tol = tol) : 
  function 'as_cholmod_sparse' not provided by package 'Matrix'

Potential Solutions:
1. update/install `irlba` and `matrix` packages from source. [github](https://github.com/satijalab/seurat/issues/8100)
+ outcome: didn't work, still get the errors
+ current version of Matrix (1.6-4), irlba (2.3.5.1), Seurat (4.3.0.1)

2. Downgrade Matrix to 1.5-3 or 1.6-1.1
+ outcomes: didn't work, breaks compatibility with SeuratObject and Signac

Another issue we encountered while updating/downgrading the Matrix Packages; Compilation issues:
> ld: warning: directory not found for option '-L/usr/local/gfortran/lib/gcc/x86_64-apple-darwin18/8.2.0'
ld: warning: directory not found for option '-L/usr/local/gfortran/lib'
ld: library not found for -lgfortran
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [Matrix.so] Error 1
ERROR: compilation failed for package ‘Matrix’

+ working [solution](https://medium.com/biosyntax/following-up-library-dependency-when-compiling-r-packages-89f191b9f227)
--------
After feature selection: 157203 Variable features

Due to the sparsity of scATAC-seq data, standard PCA analysis is not done. We would see a lot of intercell similarity due to the high frequency of 0's due to either non-accessible sites or not sampled sites. Latent Semantic Indexing (LSI) followed by SVD will find the most valuable info across samples and represent it in lower dimensional space. Thus, the dimensions are reduced and can be further visualized using UMAP or t-SNE (see next section)

Correlation between depth and reduced dimension component plot shows that the --first-- LSI component is highly correlated with total number of counts for the cell. Downstream analysis will exclude this component and can be subset away with the `dim` parameter eg. `RunUMAP(..., dims = 2:30`.

note: 
1. Latent Semantic Indexing (lsi): From NLP, used to assess document similarity based on word counts. Adapting this NLP technique to single-cell data, it treats each sample as a 'document' and each peak/region as 'words'. 
2. Term Frequency-Inverse Document Frequency Normalization (TF-IDF) is the metric that calculates the term frequency by depth normalization per cell, which is then normalized by the inverse of the document frequency.
Thus each peak or 'word' is assigned an importance value. 


## Plot UMAP with clusters
```{r Plot the UMAP with clusters}
# Runs the Uniform Manifold Approximation and Projection (UMAP) dimensional reduction technique on the `lsi` reduced data
brain <- RunUMAP(
  object = brain,
  reduction = 'lsi',
  dims = 2:30
)

# Nearest-neighbor graph construction
brain <- FindNeighbors(
  object = brain,
  reduction = 'lsi',
  dims = 2:30
)
# Cluster Determination
brain <- FindClusters(
  object = brain,
  algorithm = 3, # SLM Algorithm
  resolution = 1.2,
  verbose = FALSE
)

# plot UMAP
DimPlot(object = brain, label = TRUE) + NoLegend()

SeuratObject::Graphs(brain)
SeuratObject::Neighbors(brain)
SeuratObject::Assays(brain)
SeuratObject::Reductions(brain)
brain$seurat_clusters
```
# Create a gene activity matrix

We can quantify the activity by calculating the accessibility within each gene region, namely their respective promoter and gene body. This is simply done by summing the number of fragments that intersect the coordinates of the gene body and promoter region. Note, the function `Signac::GeneActivity()` extends the capture region for each TSS upstream by 2kb (default) to capture the promoter. The capture region at the TTS is not extended. 

```{r gene activity matrix}
# compute gene activities matrix (computes counts per cell in gene body and promoter region)
gene.activities <- GeneActivity(brain) 

# add the gene activity matrix to the Seurat object as a new assay
brain[['RNA']] <- CreateAssayObject(counts = gene.activities)
brain <- NormalizeData(
  object = brain,
  assay = 'RNA',
  normalization.method = 'LogNormalize',
  scale.factor = median(brain$nCount_RNA)
)


```
Runtime: ~ 7 mins

## Visualize gene activity at select marker genes
Now we use the gene activity matrix as a proxy of gene activity. This is similar to plotting gene expression for a given set of marker genes, but note the scATAC-seq data is much more sparse, and that accessibility at a gene locus doesn't mean the gene is actually expressed.

```{r gene activity plot}
# switch the default assay from 'peaks' to 'RNA'
DefaultAssay(brain) <- 'RNA'

# plot UMAPs with gene activity at indicated genes (set via features param)
FeaturePlot(
  object = brain,
  features = c('Sst','Pvalb',"Gad2","Neurod6","Rorb","Syt6"),
  pt.size = 0.1,
  max.cutoff = 'q95',
  ncol = 3
)
```

# Integration with scRNA-seq data
We can attempt to assign labels of cell types or cell states to the scATAC-seq clusters we generated earlier by using scRNA-seq data form the same biological system. This cross modality mapping can be done by Seurat using the reference labels from scRNA-seq data and project them into the scATAC-seq clusters.

The normal brain scRNA-seq provided by the Allen institute will be used for this integration: 

We will download the pre-processed Seurat object and load it into R. Further info regarding the code used to construct the object form the [raw](http://celltypes.brain-map.org/api/v2/well_known_file_download/694413985) can be found [here](https://github.com/satijalab/Integration2019/blob/master/preprocessing_scripts/allen_brain.R)

```{bash download allen brain scRNA-seq seurat object; echo = FALSE}
wget -P 'data/vignette_data/allen_brain_reference' https://signac-objects.s3.amazonaws.com/allen_brain.rds
```

```{r integration with scRNA-seq}
# Load the pre-processed scRNA-seq data
allen_rna <- readRDS("data/vignette_data/allen_brain_reference/allen_brain.rds")
allen_rna <- UpdateSeuratObject(allen_rna)
allen_rna <- FindVariableFeatures(
  object = allen_rna,
  nfeatures = 5000
)
# find a set of anchors between a reference and query dataset
transfer.anchors <- FindTransferAnchors(
  reference = allen_rna,
  query = brain,
  reduction = 'cca', # canonical correlation analysis
  dims = 1:40
)

# transfer anchors across single-cell datasets (scRNA-seq -> scATAC-seq)
predicted.labels <- TransferData(
  anchorset = transfer.anchors,
  refdata = allen_rna$subclass,
  weight.reduction = brain[['lsi']],
  dims = 2:30 # excluded first lsi component due to high correlation w/ counts
)

brain <- AddMetaData(object = brain, metadata = predicted.labels)
```

When running the FindTransferAnchors() function, the reduction param selected is 'cca', which is the canonical correlation analysis. Other dimension reduction algorithms can be chosen for finding anchors.

What is CCA and why is it used? Stuart et al. implemented previous for integrating single cell data. See his paper [here](https://doi.org/10.1016/j.cell.2019.05.031)

## Plot UMAP with transferred labels from Allen brain scRNA-seq data
```{r umap with scRNA-seq label transfers}
plot1 <- DimPlot(allen_rna, group.by = 'subclass', label = TRUE, repel = TRUE) + NoLegend() + ggtitle('scRNA-seq')
plot2 <- DimPlot(brain, group.by = 'predicted.id', label = TRUE, repel = TRUE) + NoLegend() + ggtitle('scATAC-seq')
plot1 + plot2
```

We can see that most of the scATAC-seq clusters are now annotated with classification labels from scRNA-seq. A few on the left are not labelled. We can further transfer labels with their most likely predictions and assigned them at a per-cell basis.

```{r label predictions}
# Here are the 23 scATAC-seq clusters (only numbered)
levels(brain)

# access the barcodes
Idents(brain) %>% head()

# replace each label with its most likely prediction
for(i in levels(brain)) {
  cells_to_reid <- WhichCells(brain, idents = i) # identify cells matching certain criteria
  newid <- names(which.max(table(brain$predicted.id[cells_to_reid]))) # choose the most common label in that cluster
  Idents(brain, cells = cells_to_reid) <- newid
}
# now the numbered clusters have been renamed with their likely predictions
levels(brain)

```

# Differentialy Accessible Peak Analysis (slow)

We can identify differentially accessible (DA) peaks by comparing clusters. Two statistical approaches are suggested:
1. Wilcoxon rank sum test via `presto` package. Can be run fast on a Seurat object
2. Logistic regression (LR), with total number of fragments as a latent variable to control for effects of sequencing depth. (used in tutorial)

Here we identify DA peaks between interneurons found in two different cortical layers, layer 2/3 Interneurons vs layer 4/5/6 Interneurons.

```{r find differentially accessible peaks, eval = FALSE}
#switch back to working with peaks instead of gene activities
DefaultAssay(brain) <- 'peaks'

da_peaks <- FindMarkers(
  object = brain,
  ident.1 = c("L2/3 IT"), 
  ident.2 = c("L4", "L5 IT", "L6 IT"),
  test.use = 'LR',
  latent.vars = 'nCount_peaks'
)
```
> This took 1 hour to run

```{r DA peak stats}
head(da_peaks)

# total peaks:
da_peaks %>%
  nrow() # 39004

# total DA peaks (padj < 0.05)
da_peaks %>% 
  filter(p_val_adj < 0.05) %>%
  nrow() # 9405
```

## plot selected DA peak of interest
```{r violin and feature plot for select DA peak}
# choose the first DA peak
rownames(da_peaks)[1]

# violin plot for the first DA peak
plot3 <- VlnPlot(
  object = brain,
  features = rownames(da_peaks)[1],
  pt.size = 0.1,
  idents = c("L4","L5 IT","L6 IT","L2/3 IT")
)

# umap showing accessibility of that peak
plot4 <- FeaturePlot(
  object = brain,
  features = rownames(da_peaks)[1],
  pt.size = 0.1,
  max.cutoff = 'q95'
)

# plots
(plot1 + plot2)/(plot3 | plot4)

(plot2)/(plot3 | plot4)

plot5 <- FeaturePlot(
  object = brain,
  features = rownames(da_peaks)[1],
  pt.size = 0.1,
  max.cutoff = 'q10'
)

# Compare q95 vs q10 as max cut off
plot4 | plot5

FeaturePlot(
  object = brain,
  features = rownames(da_peaks)[1],
  pt.size = 0.1,
  min.cutoff = 'q95'
)
```

## Differential Accessibility by fold change (fast, but not statistically rigorous)
Quick way to find DA peaks between two groups by comparing fold changes, but does not control for latent variables, such as the total sequencing depth between cells, nor does it allow for filtering by p value since no statistical tests are done.

```{r fold changes for DA peaks}
# calculate fold change between L2/3 IT vs L4/L5/L6
fc <- FoldChange(brain, 
                 ident.1 = c("L2/3 IT"),
                 ident.2 = c("L4", "L5 IT", "L6 IT")
)

# order by fold change
fc <- fc[order(fc$avg_log2FC, decreasing = TRUE), ]
head(fc)

# Can also use the fold changes calculated form LR DA peak analysis:
  # filter top DA peaks open in Layers 2/3 compared to Layer4/5/6
open_l23 <- rownames(da_peaks[da_peaks$avg_log2FC > 3, ])

  # filter top DA peaks open in Layer4/5/6 compared to Layers 2/3
open_l456 <- rownames(da_peaks[da_peaks$avg_log2FC < 3, ])

  # map DA peaks to closest gene
closest_l23 <- ClosestFeature(brain, open_l23)
closest_l456 <- ClosestFeature(brain, open_l456)
head(closest_l23)
head(closest_l456)

```
After mapping the DA peaks to genes, we can conduct functional enrichment analysis to better understand the pathways that are enriched between the clusters (layer 2/3 vs layer 4/5/6 IT).

# Plotting genomic regions
The pseudobulk accessibility signal at Genomic regions can be plotted using the `CoveragePlot()` function with additional annotations that help interpretation of chromatin accessibility at the genomic level.
```{r plot genomic regions}
# set plotting order
levels(brain) <- c("L2/3 IT","L4","L5 IT","L5 PT","L6 CT", "L6 IT","NP","Sst","Pvalb","Vip","Lamp5","Meis2","Oligo","Astro","Endo","VLMC","Macrophage")

# plot genomic region at specific DA peak interval
CoveragePlot(
  object = brain,
  region = rownames(da_peaks)[1],
  extend.upstream = 40000,
  extend.downstream = 20000
)

# plot genomic region at Neurod6 and Gad6 locus

CoveragePlot(
  object = brain,
  region = c("Neurod6", "Gad2"),
  extend.upstream = 1000,
  extend.downstream = 1000,
  ncol = 1
)

# plot the gene with top DA peak in L2/3, and top DA peak in L4/5/6
CoveragePlot(
  object = brain,
  region = c("Fam154a", "Fam19a5"),
  extend.upstream = 1000,
  extend.downstream = 1000,
  ncol = 1
)
```


